In [1]:
import numpy


In [2]:
import numpy as np
np.random.seed(42)

matrix = np.random.randint(low=0, high=10, size=(4, 4))
print(matrix)


[[6 3 7 4]
 [6 9 2 6]
 [7 4 3 7]
 [7 2 5 4]]


In [3]:
import numpy as np

vector1 = np.random.randint(low=0, high=10, size=(4, 1))
vector2 = np.random.randint(low=0, high=10, size=(4, 1))

print("Vector 1:")
print(vector1)

print("Vector 2:")
print(vector2)


Vector 1:
[[1]
 [7]
 [5]
 [1]]
Vector 2:
[[4]
 [0]
 [9]
 [5]]


In [4]:
result1 = np.dot(matrix, vector1)
result2 = np.dot(matrix, vector2)

print("Result 1:")
print(result1)

print("Result 2:")
print(result2)


Result 1:
[[66]
 [85]
 [57]
 [50]]
Result 2:
[[107]
 [ 72]
 [ 90]
 [ 93]]


In [5]:
concatenated_matrix = np.concatenate((vector1, vector2), axis=1)
result = np.dot(matrix, concatenated_matrix)

print("Result:")
print(result)


Result:
[[ 66 107]
 [ 85  72]
 [ 57  90]
 [ 50  93]]


In [18]:
batch_size = 8
layers = [4,2,4,2,1]
#activations[layer][batchsize][j]
activation_1 = np.random.uniform(low=-1, high=1, size=(batch_size, 4))
delta_2 = np.random.uniform(low=-1, high=1, size=(batch_size, 2))

print(f"")
print(activation_1.shape)
print(activation_1)
print(delta_2.shape)
print(delta_2)
print((np.dot(activation_1.T,delta_2)).shape)

print(f'first row of activation_1: {activation_1[0]}')
print(f'first row of delta_2: {delta_2[0]}')
print(f'outer product of first rows:\n{np.outer(activation_1[0],delta_2[0])}')



'''
sample = 0
j = 1
k = 0

activations_1[0][0] = -0.03666
delta_2[0][1] = 0.999
grad_weight[1][0] = 0.10029716
'''

import numpy as np

# Example activations (a) and deltas (δ) matrices
a = np.random.rand(8, 2) # replace with your actual activations matrix
δ = np.random.rand(8, 4) # replace with your actual deltas matrix

# Initialize an empty list to hold the gradients for each sample
gradients = []

# Compute the gradient for each sample
for i in range(8):
    # Compute the outer product of activations and deltas for the current sample
    gradient = np.outer(a[i], δ[i])
    # Add the gradient to our list of gradients
    gradients.append(gradient)]

# Stack the gradients to get a 3D matrix of shape (8, 2, 4)
gradients_matrix = np.array(gradients)


(8, 4)
[[-0.03666602 -0.24402482  0.41016873 -0.50255169]
 [-0.33949494 -0.13109631 -0.49263565 -0.18959583]
 [ 0.14294605  0.48194772  0.53438971  0.6455807 ]
 [ 0.48842331  0.36207885 -0.52498706 -0.19955421]
 [-0.04457022 -0.83421736  0.05674024 -0.12732834]
 [ 0.6042185   0.95580134  0.11201008 -0.35462722]
 [-0.91319843  0.84928666  0.83823143 -0.49401966]
 [ 0.39082225 -0.84913091 -0.66756914 -0.56638179]]
(8, 2)
[[-0.41101211  0.99166275]
 [ 0.39385011 -0.23159634]
 [ 0.47420141  0.83050868]
 [ 0.91740481 -0.8842722 ]
 [-0.2109564  -0.78648787]
 [-0.32867987 -0.66064019]
 [ 0.29369672 -0.22346323]
 [-0.54121051 -0.46812415]]
(4, 2)
first row of activation_1: [-0.03666602 -0.24402482  0.41016873 -0.50255169]
first row of delta_2: [-0.41101211  0.99166275]
outer product of first rows:
[[ 0.01507018 -0.03636033]
 [ 0.10029716 -0.24199032]
 [-0.16858431  0.40674905]
 [ 0.20655483 -0.49836179]]


In [24]:
import numpy as np

array = np.zeros((4, 2))
t = []
for j in range(layers[0]):
    for k in range(layers[1]):
        a_minus1_k = activation_1[0][j]
        d_j = delta_2[0][k]
        array[j][k] = a_minus1_k * d_j

print(f'weight by weight: \n{array}')
print(f'by outer product: \n{np.outer(activation_1[0],delta_2[0])}')

weight by weight: 
[[ 0.01507018 -0.03636033]
 [ 0.10029716 -0.24199032]
 [-0.16858431  0.40674905]
 [ 0.20655483 -0.49836179]]
by outer product: 
[[ 0.01507018 -0.03636033]
 [ 0.10029716 -0.24199032]
 [-0.16858431  0.40674905]
 [ 0.20655483 -0.49836179]]


(8, 4)
[[-0.03666602 -0.24402482  0.41016873 -0.50255169]
 [-0.33949494 -0.13109631 -0.49263565 -0.18959583]
 [ 0.14294605  0.48194772  0.53438971  0.6455807 ]
 [ 0.48842331  0.36207885 -0.52498706 -0.19955421]
 [-0.04457022 -0.83421736  0.05674024 -0.12732834]
 [ 0.6042185   0.95580134  0.11201008 -0.35462722]
 [-0.91319843  0.84928666  0.83823143 -0.49401966]
 [ 0.39082225 -0.84913091 -0.66756914 -0.56638179]]
(8, 2)
[[-0.41101211  0.99166275]
 [ 0.39385011 -0.23159634]
 [ 0.47420141  0.83050868]
 [ 0.91740481 -0.8842722 ]
 [-0.2109564  -0.78648787]
 [-0.32867987 -0.66064019]
 [ 0.29369672 -0.22346323]
 [-0.54121051 -0.46812415]]
(4, 2)
first row of activation_1: [-0.03666602 -0.24402482  0.41016873 -0.50255169]
first row of delta_2: [-0.41101211  0.99166275]
outer product of first rows:
[[ 0.01507018 -0.03636033]
 [ 0.10029716 -0.24199032]
 [-0.16858431  0.40674905]
 [ 0.20655483 -0.49836179]]

In [27]:
gradients = np.einsum('ij,ik->ijk', activation_1, delta_2)
# This einsum call specifies that for each i (which ranges over the rows of the matrices), you want to multiply activations[i, j] with deltas[i, k] to produce gradients[i, j, k], which is exactly the outer product of the vectors for each sample.
print(gradients.shape)

(8, 4, 2)


In [50]:
import numpy as np


matrix1 = np.arange(1, 10).reshape(3, 3)
matrix2 = np.arange(1, 10).reshape(3, 3)
matrix2 = matrix2.T
print("Matrix 1:")
print(matrix1)

print("Matrix 2:")
print(matrix2)

matrix = np.zeros((3, 3))
print("Output Matrix:")
print(matrix)

for i in range(3):
        row = matrix1[i, :]
        column = matrix2[i, :]
        
        # print(row, column)
        matrix += np.outer(row, column)

print("Output Matrix filled:")
print(matrix)
print(matrix1*matrix2)

        


Matrix 1:
[[1 2 3]
 [4 5 6]
 [7 8 9]]
Matrix 2:
[[1 4 7]
 [2 5 8]
 [3 6 9]]
Output Matrix:
[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
Output Matrix filled:
[[ 48.  60.  72.]
 [ 60.  75.  90.]
 [ 72.  90. 108.]]
[[ 1  8 21]
 [ 8 25 48]
 [21 48 81]]


In [46]:
#wrong
import numpy as np

A = np.array([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]])

B = np.array([[1, 4, 7],
              [2, 5, 8],
              [3, 6, 9]])

AB = np.dot(A, B)

print(AB)


[[ 14  32  50]
 [ 32  77 122]
 [ 50 122 194]]
